syntax = "proto3";

package Swoq.Interface;

enum Result {
    RESULT_OK = 0;
    RESULT_INTERNAL_ERROR = 1;
    RESULT_UNKNOWN_USER = 2;
    RESULT_UNKNOWN_GAME_ID = 3;
    RESULT_USER_LEVEL_TOO_LOW = 4;
    RESULT_QUEST_QUEUED = 5;
    RESULT_MOVE_NOT_ALLOWED = 6;
    RESULT_NO_PROGRESS = 7;
    RESULT_UNKNOWN_ACTION = 8;
    RESULT_GAME_FINISHED = 9;
}

enum Tile {
    TILE_UNKNOWN = 0;
    TILE_EMPTY = 1;
    TILE_PLAYER = 2;
    TILE_WALL = 3;
    TILE_EXIT = 4;
}

enum DirectedAction {
    DIRECTEDACTION_MOVE_NORTH = 0;
    DIRECTEDACTION_MOVE_EAST = 1;
    DIRECTEDACTION_MOVE_SOUTH = 2;
    DIRECTEDACTION_MOVE_WEST = 3;
}

service GameService {
    rpc Start(StartRequest) returns (StartResponse);
    rpc Act(ActionRequest) returns (ActionResponse);
}

message Position {
    int32 x = 1;
    int32 y = 2;
}

message PlayerState {
    Position position = 1;
    repeated Tile surroundings = 2;
}

message State {
    int32 tick = 1;
    int32 level = 2;
    bool finished = 3;
    optional PlayerState playerState = 4;
}

message StartRequest {
    string userId = 1;
    optional int32 level = 2;
}

message StartResponse {
    Result result = 1;
    optional string gameId = 2;
    optional int32 width = 3;
    optional int32 height = 4;
    optional int32 visibilityRange = 5;
    optional State state = 6;
}

message ActionRequest {
    string gameId = 1;
    optional DirectedAction action = 2;
}

message ActionResponse {
    Result result = 1;
    optional State state = 2;
}
